File: kotlin/app/src/main/kotlin/com/hyperbabel/demo/utils/TimeUtils.kt
/*
 * Locale-aware timestamp formatting helpers, mirroring the React / Flutter
 * demos. Intentionally tiny — every chat surface uses the same pair.
 */
package com.hyperbabel.demo.utils

import java.time.Instant
import java.time.LocalDate
import java.time.LocalDateTime
import java.time.ZoneId
import java.time.format.DateTimeFormatter
import java.time.format.FormatStyle
import java.time.temporal.ChronoUnit
import java.util.Locale

private val timeFmt: DateTimeFormatter =
    DateTimeFormatter.ofLocalizedTime(FormatStyle.SHORT).withLocale(Locale.getDefault())
private val shortDateFmt: DateTimeFormatter =
    DateTimeFormatter.ofPattern("MMM d", Locale.getDefault())
private val fullDateFmt: DateTimeFormatter =
    DateTimeFormatter.ofPattern("MMM d, yyyy", Locale.getDefault())

fun formatMessageTime(iso: String?, now: LocalDateTime = LocalDateTime.now()): String {
    if (iso.isNullOrBlank()) return ""
    val instant = runCatching { Instant.parse(iso) }.getOrNull() ?: return ""
    val when_ = LocalDateTime.ofInstant(instant, ZoneId.systemDefault())
    val secs = ChronoUnit.SECONDS.between(when_, now)
    if (secs < 60) return "just now"
    val mins = ChronoUnit.MINUTES.between(when_, now)
    if (mins < 60) return "$mins min ago"
    val today = now.toLocalDate()
    val whenDate = when_.toLocalDate()
    if (whenDate == today) return when_.format(timeFmt)
    if (whenDate == today.minusDays(1)) return "Yesterday ${when_.format(timeFmt)}"
    if (whenDate.year == today.year) return "${when_.format(shortDateFmt)} ${when_.format(timeFmt)}"
    return when_.format(fullDateFmt)
}

fun formatDateSeparator(iso: String?, today: LocalDate = LocalDate.now()): String {
    if (iso.isNullOrBlank()) return ""
    val instant = runCatching { Instant.parse(iso) }.getOrNull() ?: return ""
    val whenDate = LocalDateTime.ofInstant(instant, ZoneId.systemDefault()).toLocalDate()
    if (whenDate == today) return "Today"
    if (whenDate == today.minusDays(1)) return "Yesterday"
    return whenDate.format(fullDateFmt)
}

File: javascript/src/utils/timeUtils.js
/**
 * Locale-aware timestamp formatting helpers — mirrors the React demo.
 *   - within last 60 seconds  → "just now"
 *   - within last 60 minutes  → "X min ago"
 *   - same day                → time only
 *   - yesterday               → "Yesterday HH:mm"
 *   - this year               → short date + time
 *   - older                   → year + short date
 */

const LOCALE = (typeof navigator !== 'undefined' && navigator.language) || 'en-US';

export function formatMessageTime(input, now = new Date()) {
  if (!input) return '';
  const when = input instanceof Date ? input : new Date(input);
  if (Number.isNaN(when.getTime())) return '';
  const diffSec = Math.floor((now - when) / 1000);
  if (diffSec < 60) return 'just now';
  const diffMin = Math.floor(diffSec / 60);
  if (diffMin < 60) return `${diffMin} min ago`;
  const sameDay = when.toDateString() === now.toDateString();
  const time = when.toLocaleTimeString(LOCALE, { hour: 'numeric', minute: '2-digit' });
  if (sameDay) return time;
  const yesterday = new Date(now.getFullYear(), now.getMonth(), now.getDate() - 1);
  if (when.toDateString() === yesterday.toDateString()) {
    return `Yesterday ${time}`;
  }
  if (when.getFullYear() === now.getFullYear()) {
    return `${when.toLocaleDateString(LOCALE, { month: 'short', day: 'numeric' })} ${time}`;
  }
  return when.toLocaleDateString(LOCALE, { year: 'numeric', month: 'short', day: 'numeric' });
}

export function formatDateSeparator(input, now = new Date()) {
  if (!input) return '';
  const when = input instanceof Date ? input : new Date(input);
  if (when.toDateString() === now.toDateString()) return 'Today';
  const yesterday = new Date(now.getFullYear(), now.getMonth(), now.getDate() - 1);
  if (when.toDateString() === yesterday.toDateString()) return 'Yesterday';
  return when.toLocaleDateString(LOCALE, { year: 'numeric', month: 'long', day: 'numeric' });
}
